### PR TITLE
Support cross device rename

### LINKED ISSUE #2998: helm dependency update fails after helm upgrade
Hi

I'm using Helm from Concourse pipeline I just upgraded Helm to 2.6.1 (from 2.4.2) and my pipeline doesn't work anymore.
The culprit line seems to be https://github.com/kubernetes/helm/blob/v2.6.1/pkg/downloader/manager.go#L220
```console
$ helm dependency update
Hang tight while we grab the latest from your chart repositories...
...Successfully got an update from the "ci" chart repository
...Successfully got an update from the "local" chart repository
...Successfully got an update from the "stable" chart repository
Update Complete. ⎈Happy Helming!⎈
Error: Unable to move current charts to tmp dir: rename /tmp/build/put/staging-helm/cvhub-api/charts /tmp/build/put/staging-helm/cvhub-api/tmpcharts: invalid cross-device link
```
I cannot change the path used by Concourse  but if I do the same command manually on my mac from my home directory (i.e not from /tmp) it works:
```console
% ~/Downloads/helm dependency update
Hang tight while we grab the latest from your chart repositories...
...Successfully got an update from the "ci" chart repository
...Successfully got an update from the "local" chart repository
...Successfully got an update from the "stable" chart repository
Update Complete. ⎈Happy Helming!⎈
Saving 1 charts
Downloading elasticsearch-singlenode from repo http://127.0.0.1:8879
±|develop ✘| cvhub-api
% pwd
/Users/nsitbon/github/cvhub-api/scripts/helm/cvhub-api
```
As far as I understand the error and the related code it seems that you should not use Rename rather you should use Move.

Any help would be appreciated.

Best regards.

### PR DESCRIPTION
**Pull request by ysh7**
This PR includes changes mentioned on #5038 required for merging it.

closes #2998

### CODE REVIEW

File: internal/third_party/dep/fs/fs.go [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"runtime"
+	"syscall"
+
+	"github.com/pkg/errors"
+)
+
```
Comment (mattfarina) [inline internal/third_party/dep/fs/fs.go]: Can you add a comment letting us know where this code came from. You can find an example at https://github.com/helm/helm/blob/484d43913f97292648c867b56768775a55e4bba6/internal/third_party/k8s.io/kubernetes/deployment/util/deploymentutil.go#L30-L32

**Review by mattfarina — CHANGES_REQUESTED** [review]
I didn't expect to see this code again. It's from dep which is a rewrite (with many of the same interfaces) of the same functionality I wrote for Glide.

**Comment by thomastaylor312** [comment]
@ysh7 Is there any chance you could also port this to v3? Looks like it should be a pretty simple copy and paste for the third party code and some small changes to the download manager

File: internal/third_party/dep/fs/fs.go (hunk 1 above)
Comment (ysh7) [reply internal/third_party/dep/fs/fs.go]: I added comment

**Comment by ysh7** [comment]
> @ysh7 Is there any chance you could also port this to v3? Looks like it should be a pretty simple copy and paste for the third party code and some small changes to the download manager

Sure can do that. Will raise PR.

**Review by thomastaylor312 — APPROVED** [review]
This is looking good from my perspective and testing. I'll let @mattfarina give the final approval since he started the review

**Review by mattfarina — APPROVED** [review]

**Pull request merged**